File: src/SOP8/Egovframework/Base/base-account-bll/src/main/java/egovframework/base/account/bll/process/LoadManager.java
package egovframework.base.account.bll.process;

import egovframework.base.account.ibll.request.RequestLogin;
import egovframework.base.account.ibll.request.RequestLoginKey;
import egovframework.base.account.ibll.response.ResponseLoginKey;
import egovframework.base.model.account.User;
import egovframework.base.model.common.team.RegularMember;
import egovframework.base.model.account.Grade;
import egovframework.base.model.common.Site;
import egovframework.base.account.ibll.response.LoginResult;
import egovframework.base.account.ibll.response.ResponseSite;
import egovframework.base.account.ibll.response.ResponseGradeList;
import egovframework.base.account.bll.process.login.ExternalLoginManager;
import egovframework.base.account.bll.process.login.LoginManager;
import egovframework.base.generic.Holder;
import egovframework.base.dal.account.*;
import egovframework.base.dal.common.SiteDAO;
import egovframework.base.crypto.AES256Cipher;
import egovframework.base.response.resource.ID;
import egovframework.base.account.bll.resource.*;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.security.MessageDigest;
import java.time.LocalDateTime;
import java.util.*;
import javax.crypto.*;
import javax.crypto.spec.SecretKeySpec;
import javax.crypto.spec.IvParameterSpec;

public class LoadManager {
    private static final int KeySize = 32;
    private static char[] BaseArr = null;

    private GradeDAO gradeDao;
    private UserDAO userDao;
    private SiteDAO siteDao;
    private SessionDAO sessionDao;

    public LoadManager(GradeDAO gradeDao, UserDAO userDao, SiteDAO siteDao, SessionDAO sessionDao) {
        this.gradeDao = gradeDao;
        this.userDao = userDao;
        this.siteDao = siteDao;
        this.sessionDao = sessionDao;
    }

    public ResponseLoginKey getLoginKey(RequestLoginKey data, boolean isExternalLogin) {
        ResponseLoginKey result = new ResponseLoginKey();
        Holder<String> errorMessage = new Holder<String>();
        String salt = isExternalLogin ? "abc" : getSalt(data, errorMessage);

        if (salt == null || salt.isEmpty()) {
            result.setSuccess(false);
            result.setSalt("");
            result.setExternalLogin(isExternalLogin);
            result.setMessage(errorMessage.value != null ? errorMessage.value : "unknownUser");
        } else {
            result.setSuccess(true);
            result.setSalt(salt);
            result.setLoginKey(getLoginKey(data.getNum()));
            result.setExternalLogin(isExternalLogin);
            result.setMessage("");
        }

        return result;
    }

    public static String getLoginKey(Long num) {
        return makeRandomKey(num);
    }

    public LoginResult login(RequestLogin data, String externalLoginUrl, boolean autoLogin) {
        LoginResult result;

        if (!data.getValue().isEmpty() || !data.getKey().isEmpty()) {
            try {
                String decrypted = decrypt(data.getValue(), data.getKey());
                int index = decrypted.indexOf('|');

                if (index > 0) {
                    String userId = decrypted.substring(0, index).trim();
                    String password = decrypted.substring(index + 1).trim();
                    result = login(userId, password, data.getKey(), externalLoginUrl, autoLogin);

                    if (result.isSuccess()) {
                        result.setMessage("successToLogin");
                    }
                } else {
                    result = new LoginResult(false, "invalidFormat");
                }
            } catch (Exception e) {
                result = new LoginResult(false, e.getMessage());
            }
        } else {
            result = new LoginResult(false, "emptyData");
        }

        return result;
    }

    private LoginResult login(String userId, String password, String sessionKey, String externalLoginUrl, boolean autoLogin) {
        LoginResult result;

        if (externalLoginUrl != null && !externalLoginUrl.isEmpty()) {
            ExternalLoginManager externalLoginManager = new ExternalLoginManager(gradeDao, userDao, siteDao, sessionDao);
            result = externalLoginManager.externalLogin(userId, password, externalLoginUrl, sessionKey, autoLogin);
            if (!result.isSuccess()) return result;

            if (result.getUser() == null) {
                result.setSuccess(false);
                result.setMessage("invalidUserInfo");
            }
            return result;
        } else {
            LoginManager loginManager = new LoginManager(userDao, sessionDao, gradeDao);
            return loginManager.login(userId, password, sessionKey, autoLogin);
        }
    }

    public ResponseSite getAllSites() {
        Holder<String> errorMessage = new Holder<String>();
        List<Site> sites = this.siteDao.selectSites(null, null, null, errorMessage);
        if (sites == null || sites.isEmpty()) {
            return new ResponseSite(false, "Site 정보가 없습니다.");
        }

        ResponseSite response = new ResponseSite(true, "");
        response.getSites().addAll(sites);

        if (sites.size() > 1) {
            response.setUseMultiSite(true);
        }

        return response;
    }

    public static String makeSalt() {
        int length = 50;
        String chars = "0123456789_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
        StringBuilder sb = new StringBuilder();
        Random rand = new Random();

        for (int i = 0; i < length; i++) {
            int index = rand.nextInt(chars.length());
            sb.append(chars.charAt(index));
        }

        return sb.toString();
    }

    private static String makeRandomKey(Long num) {
        if (BaseArr == null) {
            BaseArr = makeBaseArray();
        }

        int max = BaseArr.length;
        int seed = num != null ? num.hashCode() : new Random().nextInt();
        Random rand = new Random(seed);
        StringBuilder key = new StringBuilder();

        for (int i = 0; i < KeySize; i++) {
            int index = rand.nextInt(max);
            key.append(BaseArr[index]);
        }

        return key.toString();
    }

    private static char[] makeBaseArray() {
        String base = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
        return base.toCharArray();
    }

    private String getSalt(RequestLoginKey data, Holder<String> errorMessage) {
        String condition = null;

        if (data.getUserID() != null) {
            condition = String.format("%s = '%s'", User.Fields.user_id, data.getUserID());
        }
        /*else if (data.Name != null && data.Data != null && data.Mode == (int)RequestLoginKey.ModeType.Email)
        {
            RegularMember member = GetRegularMember(data.Name, data.Data, null, out strErrorMessage);

            if (member == null)
                return null;
            else
                strCondition = string.Format("{0} = {1}", User.Fields.rgl_memb_sn, member.rgl_memb_sn);
        }
        else if (data.Name != null && data.Data != null && data.Mode == (int)RequestLoginKey.ModeType.PhoneNumber)
        {
            RegularMember member = GetRegularMember(data.Name, null, EncryptString(data.Data), out strErrorMessage);

            if (member == null)
                return null;
            else
                strCondition = string.Format("{0} = {1}", User.Fields.rgl_memb_sn, member.rgl_memb_sn);
        }*/
        else
        {
            errorMessage.value = ID.get(ErrorMessage.class, "moreParameter").value();
            return null;
        }

        List<User> users = this.userDao.selectUsers(null, condition, null, errorMessage);

        if (users == null) {
            return null;
        }

        if (!users.isEmpty()) {
            return users.get(0).getPassword_salt();
        }

        errorMessage.value = ID.get(ErrorMessage.class, "unknownUser").value();
        return null;
    }

    private static String decrypt(String inputHex, String key) throws Exception {
        byte[] inputBytes = hexStringToByteArray(inputHex);
        byte[] keyBytes = key.substring(0, 32).getBytes(StandardCharsets.UTF_8);
        byte[] ivBytes = key.substring(0, 16).getBytes(StandardCharsets.UTF_8);

        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        SecretKeySpec secretKey = new SecretKeySpec(keyBytes, "AES");
        IvParameterSpec iv = new IvParameterSpec(ivBytes);
        cipher.init(Cipher.DECRYPT_MODE, secretKey, iv);

        byte[] decrypted = cipher.doFinal(inputBytes);
        return new String(decrypted, StandardCharsets.UTF_8);
    }

    private static byte[] hexStringToByteArray(String s) {
        s = s.replace(" ", "");
        int len = s.length();
        byte[] data = new byte[len / 2];

        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                    + Character.digit(s.charAt(i+1), 16));
        }

        return data;
    }

    public static String encryptString(String str) {
        if (str == null) return null;

        try {
            return AES256Cipher.encrypt(str); // Java용 AES256Cipher 정의 필요
        }
        catch (Exception e)
        {
            System.out.println(e.getMessage());
        }

        return null;
    }

    public static String decryptString(String str) {
        if (str == null) return null;

        try {
            return AES256Cipher.decrypt(str); // Java용 AES256Cipher 정의 필요
        }
        catch (Exception e)
        {
            System.out.println(e.getMessage());
        }

        return null;
    }

    public ResponseGradeList getGradeList() {
        Holder<String> errorMessage = new Holder<>();
        List<Grade> grades = this.gradeDao.selectGrades(null, null, null, errorMessage);

        if (grades == null) return new ResponseGradeList(false, "등급 목록을 가져올 수 없습니다.");

        ResponseGradeList response = new ResponseGradeList(true, "");
        response.getGrades().addAll(grades);
        return response;
    }
}

File: src/SOP8/Base/Account/Base.Account.BLL/Process/LoadManager.cs
using System;
using Base.Account.IBLL.Request;
using Base.Account.IBLL.Response;
using System.Collections.Generic;
using Base.Model.Account;
using Base.Model.Common.Team;
using Base.Model.Common;
using Response.Resource;
using dnsDapperDBUtil.DataAccessLayer.IDAL;
using System.Text;
using System.Security.Cryptography;
using System.IO;

namespace Base.Account.BLL.Process
{
    using Resource;
    using Login;

    class LoadManager
    {
        private const int KeySize = 32;
        private static char[] BaseArr = null;

        private static char[] MakeBaseArray()
        {
            char[] arr = new char[62];
            int i = 0;

            for (char ch = '0'; ch <= '9'; ch++)
            {
                arr[i++] = ch;
            }

            for (char ch = 'a'; ch <= 'z'; ch++)
            {
                arr[i++] = ch;
            }

            for (char ch = 'A'; ch <= 'Z'; ch++)
            {
                arr[i++] = ch;
            }

            return arr;
        }

        private IDataManager m_dataManager = null;

        public LoadManager(IDataManager dataManager)
        {
            m_dataManager = dataManager;
        }

        public ResponseLoginKey GetLoginKey(RequestLoginKey data, bool isExternalLogin)
        {
            ResponseLoginKey result = new ResponseLoginKey();

            string strErrorMessage = null;
            string strSalt = isExternalLogin ? "abc" : GetSalt(data, out strErrorMessage);

            if (strSalt == null || strSalt.Length == 0)
            {
                result.Success = false;
                result.Salt = "";
                result.ExternalLogin = isExternalLogin;

                if (strErrorMessage != null)
                    result.Message = strErrorMessage;
                else
                    result.Message = ID.Get<ErrorMessage>("unknownUser").Value();
            }
            else
            {
                result.Success = true;
                result.Salt = strSalt;
                result.LoginKey = GetLoginKey(data.Num);
                result.ExternalLogin = isExternalLogin;
                result.Message = "";
            }

            return result;
        }

        public static string GetLoginKey(long? num)
        {
            string strKey = MakeRandomKey(num);
            return strKey;
        }

        public LoginResult Login(RequestLogin data, string strExternalLoginUrl, bool autoLogin)
        {
            LoginResult result = null;

            if (data.Value != "" || data.Key != "")
            {
                try
                {
                    string str = Decrypt(data.Value, data.Key);

                    int nIndex = str.IndexOf('|');

                    if (nIndex > 0)
                    {
                        string strID = str.Substring(0, nIndex).Trim();
                        string strPW = str.Substring(nIndex + 1).Trim();

                        result = Login(strID, strPW, data.Key, strExternalLoginUrl, autoLogin);

                        if (result.Success)
                        {
                            result.Message = ID.Get<ErrorMessage>("successToLogin").Value();
                        }
                    }
                    else
                    {
                        result = new LoginResult();
                        result.Success = false;
                    }
                }
                catch (Exception e)
                {
                    result = new LoginResult();
                    result.Message = e.Message;
                    result.Success = false;
                }
            }
            else
            {
                result = new LoginResult();
                result.Success = false;
            }

            return result;
        }

        private LoginResult Login(string strUserID, string strPW, string strSessionKey, string strExternalLoginURL, bool autoLogin)
        {
            LoginResult result = null;

            if (strExternalLoginURL != null && strExternalLoginURL.Length > 0)
            {
                User user;
                Grade grade;
                ExternalLoginManager externalLoginManager = new ExternalLoginManager(m_dataManager);
                result = externalLoginManager.ExternalLogin(strUserID, strPW, strExternalLoginURL, strSessionKey, autoLogin, out user, out grade);

                if (result.Success == false)
                    return result;
                else if (result.User == null)
                {
                    result.Success = false;
                    result.Message = ID.Get<ErrorMessage>("invalidUserInfo").Value();
                }

                return result;
            }
            else
            {
                LoginManager loginManager = new LoginManager(m_dataManager);
                result = loginManager.Login(strUserID, strPW, strSessionKey, autoLogin);
            }

            return result;
        }

        public ResponseSite GetAllSites()
        {
            string strErrorMessage;
            IEnumerable<Site> sites = m_dataManager.GetSelect().Select<Site>(null, out strErrorMessage);

            if (sites == null)
                return new ResponseSite(false, strErrorMessage);

            ResponseSite response = new ResponseSite(true, "");
            response.Sites.AddRange(sites);

            if (response.Sites.Count == 0)
                return new ResponseSite(false, "Site 정보가 없습니다.");
            else if (response.Sites.Count > 1)
                response.UseMultiSite = true;

            return response;
        }

        public static string MakeSalt()
        {
            int length = 50;
            string strChars = "0123456789_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

            int charLength = strChars.Length;
            string strData = "";

            Random rand = new Random((int)DateTime.Now.ToBinary());

            for (int i = 0; i < length; i++)
            {
                int index = rand.Next(0, charLength - 1);
                strData += strChars[index];
            }

            return strData;
        }

        private static string MakeRandomKey(long? num)
        {
            string strKey = "";

            if (BaseArr == null)
                BaseArr = MakeBaseArray();

            int max = BaseArr.Length - 1;

            int seed = num == null ? DateTime.Now.GetHashCode() : (int)num;
            Random rand = new Random(seed);

            for (int i = 0; i < KeySize; i++)
            {
                int nIndex = rand.Next(max);
                strKey += BaseArr[nIndex];
            }

            return strKey;
        }

        private string GetSalt(RequestLoginKey data, out string strErrorMessage)
        {
            string strCondition = "";
            Dictionary<User.Fields, object> dicConditions = new Dictionary<User.Fields, object>();

            if (data.UserID != null)
                strCondition = string.Format("{0} = '{1}'", User.Fields.user_id, data.UserID);
            else if (data.Name != null && data.Data != null && data.Mode == (int)RequestLoginKey.ModeType.Email)
            {
                RegularMember member = GetRegularMember(data.Name, data.Data, null, out strErrorMessage);

                if (member == null)
                    return null;
                else
                    strCondition = string.Format("{0} = {1}", User.Fields.rgl_memb_sn, member.rgl_memb_sn);
            }
            else if (data.Name != null && data.Data != null && data.Mode == (int)RequestLoginKey.ModeType.PhoneNumber)
            {
                RegularMember member = GetRegularMember(data.Name, null, EncryptString(data.Data), out strErrorMessage);

                if (member == null)
                    return null;
                else
                    strCondition = string.Format("{0} = {1}", User.Fields.rgl_memb_sn, member.rgl_memb_sn);
            }
            else
            {
                strErrorMessage = ID.Get<ErrorMessage>("moreParameter").Value();
                return null;
            }

            IEnumerable<User> users = m_dataManager.GetSelect().Select<User>(strCondition, out strErrorMessage);

            if (users == null)
                return null;

            foreach (User user in users)
            {
                return user.password_salt;
            }

            strErrorMessage = ID.Get<ErrorMessage>("unknownUser").Value();
            return null;
        }

        private RegularMember GetRegularMember(string strName, string strEmail, string strPhoneNumber, out string strErrorMessage)
        {
            strErrorMessage = null;

            string strCondition = string.Format("{0} = '{1}'", RegularMember.Fields.memb_name, strName);

            if (strEmail != null)
                strCondition += string.Format(" and {0} = '{1}'", RegularMember.Fields.email, strEmail);
            else if (strPhoneNumber != null)
                strCondition += string.Format(" and {0} = '{1}'", RegularMember.Fields.telno, strPhoneNumber);
            else
            {
                strEmail = ID.Get<ErrorMessage>("moreParameter").Value();
                return null;
            }

            IEnumerable<RegularMember> members = m_dataManager.GetSelect().Select<RegularMember>(strCondition, out strErrorMessage);

            if (members == null)
                return null;

            foreach (RegularMember member in members)
            {
                return member;
            }

            strErrorMessage = ID.Get<ErrorMessage>("unknownID").Value();
            return null;
        }

        public ResponseGradeList GetGradeList()
        {
            string strErrorMessage;
            IEnumerable<Grade> grades = m_dataManager.GetSelect().Select<Grade>(null, out strErrorMessage);

            if (grades == null)
                return new ResponseGradeList(false, strErrorMessage);

            ResponseGradeList response = new ResponseGradeList(true, "");

            foreach (Grade grade in grades)
            {
                response.Grades.Add(grade);
            }

            return response;
        }

        public static string EncryptString(string str)
        {
            if (str == null)
                return null;

            return dnsDapperDBUtil.AES256Cipher.AES_encrypt(str);
        }

        public static string DecryptString(string str)
        {
            if (str == null)
                return null;

            return dnsDapperDBUtil.AES256Cipher.AES_decrypt(str);
        }

        private static string Decrypt(string input, string key)
        {
            byte[] inputBytes = HexStringToByteArray(input);
            byte[] keyBytes = Encoding.UTF8.GetBytes(key.Substring(0, 32));
            using (AesCryptoServiceProvider aesAlg = new AesCryptoServiceProvider())
            {
                aesAlg.Key = keyBytes;
                aesAlg.IV = Encoding.UTF8.GetBytes(key.Substring(0, 16));

                ICryptoTransform decryptor = aesAlg.CreateDecryptor(aesAlg.Key, aesAlg.IV);
                using (MemoryStream msEncrypt = new MemoryStream(inputBytes))
                {
                    using (CryptoStream csEncrypt = new CryptoStream(msEncrypt, decryptor, CryptoStreamMode.Read))
                    {
                        using (StreamReader srEncrypt = new StreamReader(csEncrypt))
                        {
                            return srEncrypt.ReadToEnd();
                        }
                    }
                }
            }
        }

        private static byte[] HexStringToByteArray(string s)
        {
            s = s.Replace(" ", "");
            byte[] buffer = new byte[s.Length / 2];
            for (int i = 0; i < s.Length; i += 2)
                buffer[i / 2] = (byte)Convert.ToByte(s.Substring(i, 2), 16);
            return buffer;
        }
    }
}
